File: ctci/1_arrays_and_strings/1_6.py
# aabcccccaaa
# a2blc5a3

def compressor(string):
    breaks = 1
    strings = []
    current_char = string[0]
    last_index_stop = 0
    for i in range(len(string)):    
        if current_char != string[i]:
            strings.append(current_char + str(i - last_index_stop))
            current_char = string[i]
            last_index_stop = i
            breaks += 1
    
    if breaks > len(string):
        return string
    strings.append(current_char+ str(len(string) - last_index_stop))
    return ''.join(strings)

print(compressor(string='aabcccccaaa'))

File: ctci/1_arrays_and_strings/1_6.js

const compressor = (string) => {
    let breaks = 1;
    let finalString = '';
    let current_char = string[0];
    let lastIndexStop = 0;

    for (let i = 0; i < string.length;i++) {
        if (current_char != string[i]) {
            finalString += current_char.concat(String(i - lastIndexStop))
            current_char = string[i]
            lastIndexStop = i 
            breaks += 1
        }
    }
    if (breaks > string.length){
        return string
    }
    else {
        finalString += current_char.concat(String(string.length - lastIndexStop))
        return finalString
    }
}

console.log(compressor('aabcccccaaa'))